### PR TITLE
Checks on delay_density

### LINKED ISSUE #89: Checks on delay_density
This issue is in reference to the function that is passed to `delay_density`. I would ideally like to resolve this issue by adding a fix to #88.

In #88, I have implemented a check on `delay_density` being a function with only argument. It could be good to constrain users in this way to avoid potential errors.

This is not really in line with other functionals as pointed out in the review comment below. I'm happy to remove this check, but it would be good to hear opinions on either side.

> My initial preference would be to allow both to be in line with what other common functions are doing (apply(), map(), etc.). Could you open an issue for this please so others can weigh in and we can take some time to consider both alternatives?

_Originally posted by @Bisaloo in https://github.com/epiverse-trace/cfr/pull/88#discussion_r1371476066_

### PR DESCRIPTION
**Pull request by pratikunterwegs**
This PR adds more checks on the `delay_density` argument to fix #89.

1. Adds the internal fn `test_fn_req_args()` to check for the number of required arguments (accounting for `...`), this is used to check that `delay_density` has only one required argument,
2. Adds the internal fn `test_fn_num_out()` to check that a function returns a numeric output consistent with evaluating a PDF/PMF, this is used to check that `delay_density` returns appropriate output types (BUT not necessarily statistically correct!): the same length as an input vector `x`, with lower bound at 0.0, but no upper bound, and no missing or NULL values. 

cc @TimTaylor for opinions.

### CODE REVIEW

**Comment by pratikunterwegs** [comment]
Rebased on to main from `dev/fixes` to allow cleaner review.

**Comment by pratikunterwegs** [comment]
Hi @Bisaloo - could you review this soon, as I'd like to merge it and the other open PRs soon.

File: R/helper_delay_density.R [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+test_fn_req_args <- function(fn, n_req_args = 1) {
+  checkmate::assert_count(n_req_args, positive = TRUE)
+  checkmate::test_function(fn) &&
+    identical(
+      length(
+        Filter(is.name, formals(fn))
+      ),
+      as.integer(n_req_args)
+    )
+}
```
Comment (Bisaloo) [inline R/helper_delay_density.R:21]: Ellipses `...` need to be special-cased because they are counted as a required argument here.

**Comment by pratikunterwegs** [comment]
Thanks, will fix this afternoon.

File: R/estimate_rolling.R [hunk 2]
```diff
@@ -66,14 +66,10 @@ cfr_rolling <- function(data,
...
     "Input data must have sequential dates with none missing or duplicated" =
-      identical(unique(diff(data$date)), 1), # use numeric 1, not integer
+      identical(unique(diff(data$date)), 1) # use numeric 1, not integer
     # this solution works when df$date is `Date`
     # this may need more thought for dates that are integers, POSIXct,
     # or other units; consider the units package
-    "`delay_density` must be a distribution density function with 1 argument
-    evaluating density at a vector of values and returning a numeric vector.
-    E.g. function(x) stats::dgamma(x = x, shape = 5, scale = 1)" =
-      checkmate::test_function(delay_density, null.ok = TRUE)
```
[the remark this answers is not in the document]
Comment (pratikunterwegs) [reply R/estimate_rolling.R:76]: Added in https://github.com/epiverse-trace/cfr/pull/94/commits/1fae77b140e2a4c31e22f1f49e87bdef4294e7ac

File: R/helper_delay_density.R (hunk 1 above)
Comment (pratikunterwegs) [reply R/helper_delay_density.R]: Added in https://github.com/epiverse-trace/cfr/pull/94/commits/9d5d0d075fcc018bbdabc80123f436eb90c9bdd9, please see if suitable. Satisfies some basic tests

Comment (Bisaloo) [reply R/helper_delay_density.R]: Another edge case:

``` r
test_fn_req_args <- function(fn, n_req_args = 1) {
  checkmate::assert_count(n_req_args, positive = TRUE)
  checkmate::test_function(fn) &&
    sum(mapply(function(x, y) {
      is.name(x) && y != "..."
    }, formals(fn), names(formals(fn)))) == n_req_args
}

test_fn_req_args(is.matrix)
#> Error in sum(mapply(function(x, y) {: invalid 'type' (list) of argument
```

<sup>Created on 2023-11-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: R/helper_delay_density.R [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+test_fn_req_args <- function(fn, n_req_args = 1) {
+  checkmate::assert_count(n_req_args, positive = TRUE)
+  checkmate::test_function(fn) &&
+    sum(mapply(function(x, y) {
+      is.name(x) && y != "..."
+    }, formals(fn), names(formals(fn)))) == n_req_args
+}
+
+#' @name delay_density_helpers
+test_fn_num_out <- function(fn, n = 10) {
```
Comment (Bisaloo) [inline R/helper_delay_density.R:41]: `n` is not used

File: R/helper_delay_density.R (hunk 1 above)
Comment (pratikunterwegs) [reply R/helper_delay_density.R]: Thanks - do you prefer for an error when a primitive is passed with no further evaluation, or to return FALSE?

Comment (Bisaloo) [reply R/helper_delay_density.R]: ?formals help suggests:

> args for a human-readable version, and as intermediary to get formals of a primitive function.

``` r
formals(args(is.matrix))
#> $x

# also works for non-primitive
formals(args(mean))
#> $x
#> 
#> 
#> $...
```

<sup>Created on 2023-11-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Comment (pratikunterwegs) [reply R/helper_delay_density.R]: Thanks both, will wrap args in formals. Just a note that I'm using `checkmate::test_function()` instead of an assert to avoid this warning when `delay_density` is `NULL` (when no correction is applied).

``` r
test_fn_req_args <- function(fn, n_req_args = 1) {
  checkmate::assert_count(n_req_args, positive = TRUE)
  checkmate::assert_function(fn, null.ok = T)
  formals <- formals(args(fn))
  no_default <- vapply(formals, is.symbol, TRUE)
  sum(no_default & names(formals) != "...") == n_req_args
}

test_fn_req_args(NULL)
#> Warning in formals(fun): argument is not a function
#> [1] FALSE

test_fn_req_args <- function(fn, n_req_args = 1) {
  checkmate::assert_count(n_req_args, positive = TRUE)
  checkmate::test_function(fn) &&
    (!is.primitive(fn)) &&
    sum(mapply(function(x, y) {
      is.name(x) && y != "..."
    }, formals(fn), names(formals(fn)))) == n_req_args
}

test_fn_req_args(NULL)
#> [1] FALSE
```

<sup>Created on 2023-11-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

I'm chaining this as so `test_fn_req_args(fn) || is.null(fn)` in the code to allow for NULLs (rather than using the checkmate argument).

Comment (pratikunterwegs) [reply R/helper_delay_density.R]: Curious whether passing `if` or `for` is another case that should be caught? Or is it safe to consider it improbable?

``` r
test_fn_req_args <- function(fn, n_req_args = 1) {
  checkmate::assert_count(n_req_args, positive = TRUE)
  checkmate::test_function(fn) &&
    sum(mapply(function(x, y) {
      is.name(x) && y != "..."
    }, formals(args(fn)), names(formals(args(fn))))) == n_req_args
}

test_fn_req_args(is.list)
#> [1] TRUE
test_fn_req_args(`+`)
#> [1] FALSE

is.function(`if`)
#> [1] TRUE
test_fn_req_args(`if`)
#> Warning in formals(fun): argument is not a function

#> Warning in formals(fun): argument is not a function
#> Error in sum(mapply(function(x, y) {: invalid 'type' (list) of argument

is.function(`for`)
#> [1] TRUE
test_fn_req_args(`for`)
#> Warning in formals(fun): argument is not a function

#> Warning in formals(fun): argument is not a function
#> Error in sum(mapply(function(x, y) {: invalid 'type' (list) of argument
```

<sup>Created on 2023-11-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Comment (pratikunterwegs) [reply R/helper_delay_density.R]: Case with `builtin`s fixed here https://github.com/epiverse-trace/cfr/pull/94/commits/346b1ff5aa589bd7096e749100e36d4e9213f30b. Still errors on `special`s.

**Comment by pratikunterwegs** [comment]
Merging this PR in about an hour, unless there is any further feedback @Bisaloo and @TimTaylor?

File: R/helper_delay_density.R [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+#'
+#' @return A logical for whether the function `fn` fulfils conditions specified
+#' in the respective checks.
+#' @keywords internal
+test_fn_req_args <- function(fn, n_req_args = 1) {
+  checkmate::assert_count(n_req_args, positive = TRUE)
+  checkmate::test_function(fn) &&
+    sum(mapply(function(x, y) {
+      is.name(x) && y != "..."
+    }, formals(args(fn)), names(formals(args(fn))))) == n_req_args
```
Comment (Bisaloo) [inline R/helper_delay_density.R:37]: Please add a comment indicating why we have to go through `formals(args())` rather than `formals()` directly. It's easy to forget.

**Review by Bisaloo — APPROVED** [review]
